eio/tls_eio.ml

```diff
@@ -5,198 +5,143 @@ exception Tls_failure of Tls.Engine.failure
 
 module Raw = struct
 
-  (* We could replace [`Eof] with [`Error End_of_file] and then use
-     a regular [result] type here. *)
   type t = {
     flow           : Flow.two_way ;
-    mutable state  : [ `Active of Tls.Engine.state
-                     | `Eof
-                     | `Error of exn ] ;
-    mutable linger : Cstruct.t option ;
+    mutable tls    : Tls.Engine.state ;
+    mutable linger : Cstruct.t ;
     recv_buf       : Cstruct.t ;
+    mutex          : Eio.Mutex.t;
   }
 
-  let read_t t cs =
-    try Flow.read t.flow cs
-    with
-    | End_of_file as ex ->
-      t.state <- `Eof;
+  let write_t t cs = Flow.copy (Flow.cstruct_source [cs]) t.flow
+
+  (* A TLS 'read' requires both reading and writing to an underlying
+     flow object - t.flow; therefore we use mutex to maintain the
+     integrity of 'read' from multiple fibres doing interleaving
+     read/write operations on the same underlying flow object. *)
+  let read_t t =
+    try
+      Eio.Mutex.lock t.mutex ;
+      let got = Flow.single_read t.flow t.recv_buf in
+      let data = Cstruct.sub t.recv_buf 0 got in
+      match Tls.Engine.handle_tls t.tls data with
+      | Ok ( state, `Response resp, `Data application_data ) ->
+        begin match state with
+          | `Ok tls ->
+            t.tls <- tls ;
+            Option.iter (write_t t) resp ;
+            Eio.Mutex.unlock t.mutex ;
+            application_data
+          | `Eof -> raise End_of_file ;
+          | `Alert a -> raise (Tls_alert a)
+        end
+      | Error (failure, `Response resp) ->
+        write_t t resp ;
+        raise (Tls_failure failure)
+    with 
+    | ( End_of_file | Tls_alert _ | Tls_failure _ ) as ex ->
```

Comment:
I am not sure but it seems when handling exceptions in effects/eio, we need to be more explicit as to what we are handling. At one time exceptions such as  this `https://github.com/ocaml/ocaml/blob/trunk/stdlib/effect.ml#L59` was being trapped in here. Perhaps this should go in the `eio` error handling guidance ?